Source/EventAnalysis/KTCollinearTrackClustering.cc

```diff
@@ -0,0 +1,450 @@
+
+                // Set up new data object
+                Nymph::KTDataPtr data( new Nymph::KTData() );
+                KTProcessedTrackData& newTrack = data->Of< KTProcessedTrackData >();
+                newTrack.SetComponent( component );
+
+                KTDEBUG(tclog, "Initalized new track");
+
+                TrackSetCIt trackIt = compIt->begin();
+
+                // Loop through tracks in the cluster
+                for( int i = 0; i < fCluster.size(); ++i )
+                {
+                    KTDEBUG(tclog, "Advancing to position " << fCluster[i] );
+                    // Get this track from the iterator
+
+                    if( i == 0 )
+                    {
+                        std::advance( trackIt, fCluster[i] );
+                    }
+                    else
+                    {
+                        std::advance( trackIt, fCluster[i] - fCluster[i-1] );
+                    }
+
+                    // Assign start and end time/frequency
+
+                    if( i == 0 || trackIt->GetStartTimeInRunC() < newTrack.GetStartTimeInRunC() )
+                    {
+                        newTrack.SetStartTimeInRunC( trackIt->GetStartTimeInRunC() );
+                    }
+                    if( i == 0 || trackIt->GetEndTimeInRunC() > newTrack.GetEndTimeInRunC() )
+                    {
+                        newTrack.SetEndTimeInRunC( trackIt->GetEndTimeInRunC() );
+                    }
+                    if( i == 0 || trackIt->GetStartFrequency() < newTrack.GetStartFrequency() )
+                    {
+                        newTrack.SetStartFrequency( trackIt->GetStartFrequency() );
+                    }
+                    if( i == 0 || trackIt->GetEndFrequency() > newTrack.GetEndFrequency() )
+                    {
+                        newTrack.SetEndFrequency( trackIt->GetEndFrequency() );
+                    }
+
+                    KTDEBUG(tclog, "Assigned time and frequency");
+
+                    // Set tracks in this cluster to GROUPED
+                    // They will no longer be considered for future clusters
+
+                    fGroupingStatuses[fCluster[i]] = fGROUPED;
+
+                    KTDEBUG(tclog, "Assigned GROUPED status");
+                }
+
+                // Process & emit new track
+
+                KTINFO(tclog, "Set clustered tracks to GROUPED. Now processing new track");
+                ProcessNewTrack( newTrack );
+
+                KTDEBUG(tclog, "Emitting track signal");
+                fTrackSignal( data );
+
+                // If there are remaining UNGROUPED tracks, the while loop will repeat
+                // If all tracks are GROUPED, we will exit this loop
+            }
+
+
+            ++component;
+        } // for loop over components
+
+        KTINFO(tclog, "Collinear track clustering finished. Emitting done signal");
+        fDoneSignal();
+
+        return true;
+    }
+
+    void KTCollinearTrackClustering::FindCluster()
```

Comment:
1. Instead of using `pow()`, pre-calculate what's being raised to a power if necessary, and then multiply it by itself.  It's a lot faster than `pow()`.

1. When you divide by `nUngrouped` it looks like you want floating point division.  In that case it'd be good to explicitly cast that to a double.